Add GL Intel performance counters support

djdeath opened the pull request:
## Description

Intel published a performance query extension support defined back in
2013. This is available on Windows & Linux (in the Mesa driver). This
should provide the same types of counters as the MDAPI backend for
DX11.

Frameretrace [1] (a fork/branch of Apitrace) uses the same extension.

[1]: https://github.com/janesma/apitrace/wiki/screen-shots

djdeath committed:
Add GL Intel performance counters support

Intel published a performance query extension support defined back in
2013. This is available on Windows & Linux (in the Mesa driver). This
should provide the same types of counters as the MDAPI backend for
DX11.

Frameretrace [1] (a fork/branch of Apitrace) uses the same extension.

v2: Fix build without OpenGL
    Simplify logic to enable counters
    Warn about non enabled counters on Linux/Mesa
    Generate counter Uuid

v3: Turn asserts into errors
    Don't load perf entry points manually

v4: More clang-format

v5: Fix some Windows conversion warnings

v6: Fix errors on Windows where the driver reports an error on
    glGetPerfQueryInfoINTEL as a mean to say that the queryId cannot
    be used through the extension

v7: clang-format

v8: Initialize variable passed by pointers to GL entry points

v9: Only try to use the INTEL_performance_query on Mesa, experience
    shows the Intel Windows driver doesn't report anything useful.

[1]: https://github.com/janesma/apitrace/wiki/screen-shots

dos1 commented:
FYI: I had to remove the `m_sampleIndex` variable from `intel_gl_counter.[cpp,h]` for it to build as it warned about unused variable and had `Werror` enabled.

djdeath commented:
Thanks, I didn't see is issue locally. Will update the PR.

djdeath commented:
If you're running this on Mesa, we currently don't have a way to nicely enable performance counters for unprivileged users (working on it though).
You'll need to run : 
```
sudo sysctl dev.i915.perf_stream_paranoid=0
```
Otherwise you'll miss out on interesting counters.

baldurk commented on renderdoc/driver/gl/gl_hooks.cpp:
```diff
@@ -177,6 +177,22 @@ void GLDispatchTable::PopulateWithCallback(PlatformGetProcAddr lookupFunc)
   }
 
   ForEachSupported(HookFunc);
+  GL.glBeginPerfQueryINTEL = (PFNGLBEGINPERFQUERYINTELPROC)lookupFunc("glBeginPerfQueryINTEL");
```
These manual lookups shouldn't be added, the macro `ForEachSupported` is auto-generated. This would also cause problems because the `Unsupported` macros contain the intel perf query entry points. See the top of `gl_dispatch_table_defs.h` for more details but basically you run the `gen_dispatch_table.py` file and the output is the newly generated `gl_dispatch_table_defs.h`.

This will mean exposing wrappers for these functions in `WrappedOpenGL` in `gl_query_funcs.cpp`, but for these functions it should be safe to just have empty non-serialising non-intercepting implementations, similar to most of the `glGet*` functions. They'll just get passed through without being captured if the application tries to use them (though it shouldn't as we don't whitelist the extension).

djdeath replied:
So you would like to have those intel entry points added to the supported ones?
I'm having a hard time getting this to work...
If I add the chunks to gl_common.h and the same case as glGet* in gl_driver.cpp that leads to asserts.
Returning true to skip them leads to other chunks being messed up somehow.

baldurk replied:
Can you give a bit more detail about what went wrong? Which asserts did you hit? One problem I can see is that you didn't add the intel functions at the end of the `GLChunk` enum, which would lead to binary incompatibility. If you loaded a capture made prior to your changes then I would expect that to break. Any `ContextConfiguration` chunk would instead hit the case for `glBeginPerfQueryINTEL` because that's what the enum value corresponds to now, and after that point the serialisation would be desynced and everything would blow up (hopefully it wouldn't crash but there's not necessarily a guarantee in all cases).

If you captured and replayed on a build with the changes then I don't see anything in principle that should break, it looks fine to me.

djdeath replied:
Ah that might be the problem then. Is there a version number I should bump so that the user is warned that traces are incompatible?

baldurk replied:
There is, but backwards compatibility should only be broken when absolutely necessary. In this case you can add the enums to the end of the list instead of adding them before existing chunk types. There won't even be need for any upgrade/version-compatibility code as the chunks won't actually be emitted, it will be cross-compatible for free.

djdeath replied:
Thanks, got it.

baldurk commented on renderdoc/driver/gl/gl_replay.cpp:
```diff
@@ -254,32 +257,44 @@ void GLReplay::SetReplayData(GLWindowingData data)
   }
 
-  if(isMesa)
+  if(isMesa && m_DriverInfo.vendor != GPUVendor::Intel)
```
I'd prefer to structure this something like:

```
if(isMesa)
{
  // handle any mesa counters (intel only for now), or else skip
}
else
{
  // handle any proprietary counters (amd only for now), or else skip.
}
```

Since it feels clearer to me to consider the split between proprietary/mesa drivers as the 'primary' one, then check different IHVs within each branch. If e.g. AMD mesa counters come along later down the line this series of ifs becomes a bit harder to read.

If I understand right this extension is available on the non-mesa windows driver too which will mean a duplicated line in each branch, but I'd rather have that minor duplication and be more clear about the code flow.

djdeath replied:
Correct, the Mesa driver needs to provide pretty close feature parity on the performance counter side because we need to support Intel GPA (https://software.intel.com/en-us/gpa)

baldurk commented on renderdoc/driver/ihv/intel/intel_gl_counters.cpp:
```diff
@@ -0,0 +1,282 @@
+{
+  vector<GPUCounter> counters;
+
+  for(const auto &c : m_Counters)
```
Avoid `auto` please unless the type is quite painful e.g. stl iterators or lambdas.

(There are a couple of other instances, I haven't comment on each to avoid spam).

baldurk commented on renderdoc/driver/ihv/intel/intel_gl_counters.cpp:
```diff
@@ -0,0 +1,282 @@
+    case GL_PERFQUERY_COUNTER_DATA_DOUBLE_INTEL: return CompType::Double;
+    case GL_PERFQUERY_COUNTER_DATA_BOOL32_INTEL: return CompType::UInt;
+    default:
+      RDCASSERT(false && "Wrong type");    // unreachable
```
Asserts should be for checking conditions, this is an error message so I'd prefer something like `RDCERR("Wrong counter data type: %u", glDataType);`

baldurk commented on renderdoc/driver/ihv/intel/intel_gl_counters.cpp:
```diff
@@ -0,0 +1,282 @@
+  GL.glGetIntegerv(eGL_PERFQUERY_COUNTER_NAME_LENGTH_MAX_INTEL, &len);
+  counter.desc.name.resize(len);
+  GL.glGetIntegerv(eGL_PERFQUERY_COUNTER_DESC_LENGTH_MAX_INTEL, &len);
+  counter.desc.description.resize(100);
```
This `100` should be `len` I think?

djdeath replied:
Correct.

baldurk commented on renderdoc/driver/ihv/intel/intel_gl_counters.cpp:
```diff
@@ -0,0 +1,282 @@
+
+  counter.desc.resultType = glToRdcCounterType(counter.dataType);
+
+  m_Counters.push_back(counter);
```
`counter.desc.uuid` isn't filled out here, which is used to identify a counter even if the enumerated order changes slightly. Without the uuid the counterset can't be saved and loaded from disk. It should just be something that has a high chance of being identical between runs and driver changes, and ideally even different hardware with overlapping sets of counters available. The UUID doesn't have to be a full 128 bits of identification, you can just pick a uuid yourself and set one word if that's all you need.

baldurk commented on renderdoc/driver/ihv/intel/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+    intel_gl_counters.cpp
+    intel_gl_counters.h)
+
+set(include_dirs ${RDOC_INCLUDES})
```
The docs CI build inadvertently found an issue here - when the GL/GLES drivers are disabled, this code is still built in and fails to compile because it references `GL` that is declared in headers but never defined.

I think there are two solutions: either make this GL code independent of the GL driver so it fetches its function pointers itself, or probably better just to make sure the `intel_gl_counters.cpp` doesn't get included if `ENABLE_GL` and `ENABLE_GLES` are both off in cmake.

djdeath replied:
Done.

baldurk reviewed (changes requested):
Thanks for this, I've made comments on some details-level things but overall it looks good to me. I've mentioned the linux CI failure (it's to do with building with GL disabled), but have a look at the windows CI as there are a couple of warnings there to fix.

Do you know what the `GL_INTEL_performance_queries` extension is? It seems to be supported on some windows systems judging by [opengl.gpuinfo.org](http://opengl.gpuinfo.org/listreports.php?extension=GL_INTEL_performance_queries) but there's no extension spec for it. I'm guessing it's an unreleased precursor and can be ignored.

As an optional extra improvement with the sysctl mentioned above - is that something we could detect within the code and suggest to the user? i.e. something like `if(dev.i915.perf_stream_paranoid == 1) { RDCWARN("Not all counters available, run 'sudo sysctl dev.i915.perf_stream_paranoid=0' to enable more counters!"); }` ? Not required but might be nice to let people know.

djdeath commented:
Thanks, I'll look into addressing all your comments asap.
Just wanted to point you to the extension quickly : https://www.khronos.org/registry/OpenGL/extensions/INTEL/INTEL_performance_query.txt (did I misspell queries/query somewhere?)

baldurk commented:
No you didn't misspell anything that I saw, I mean that there are [reports from windows drivers](http://opengl.gpuinfo.org/listreports.php?extension=GL_INTEL_performance_queries) - for example [this one](http://opengl.gpuinfo.org/displayreport.php?id=3122) where an extension `GL_INTEL_performance_queries` is reported.

djdeath commented:
I have no idea, I'll check with the Windows people. Thanks for pointing this out.

djdeath commented:
So apparently that is the same extension, there was just a mistake in the name.
Some drivers ship both : http://opengl.gpuinfo.org/displayreport.php?id=2065

baldurk commented:
OK then, it's probably safe to ignore the extension then as it seemed to only be reported in older drivers.

djdeath commented:
Updated without the removal of manual hooks for intel perf query entry points.

djdeath commented:
This is my failed attempt at removing the manual hookups : https://github.com/djdeath/renderdoc/commit/049dbd220bad64991567151b131cac2211be84bf
Help welcome!

baldurk commented on renderdoc/driver/ihv/intel/intel_gl_counters.cpp:
```diff
@@ -0,0 +1,309 @@
+          break;
+        }
+        default:
+          RDCASSERT(false && "Wrong type");    // unreachable
```
Still a couple of cases of (ab)using asserts as error messages here, FYI - see `IntelGlCounters::CounterPass` as well.

djdeath replied:
Oops, fixed.

djdeath commented:
Hopefully it's all green this time.

baldurk commented:
I'll test this next week once the CI is all happy and merge it in.

baldurk commented:
I found a problem when testing on windows. The first query enumerated is 258, and the call to `glGetPerfQueryInfoINTEL` to get the number of counters fires this error:

`'Error has been generated. GL error GL_INVALID_VALUE in GetPerfQueryInfoINTEL: (ID: 1585457696) Generic error'`

And then you end up with `nCounters` being uninitialised. On a debug build at least this leads to iterating up to 0xcccccccc or so and looping forever (and each `glGetPerfQueryInfoINTEL` for the counters then prints an error, leading to log spam).

I think at least any variables fetched by `glGet*INTEL` should be initialised to 0 if the function might not write to them, and also it'd be good to understand what's going wrong here.

baldurk reviewed (changes requested):
FYI #1229 introduced a minor conflict, you'll need to update the `GLChunk` enum to put yours at the new end, and regenerate `gl_dispatch_table_defs.h`. I'm also going to set this back to 'request changes' since there's the above bug on windows still to address.

baldurk commented:
Let me know if you want to continue this PR. I know the holidays ate up a bunch of time so I'm not in a rush here, but I'd rather not have it hanging around in limbo indefinitely.

djdeath commented:
Very sorry about that, I've just been swamped with other things and didn't find time to update it.

djdeath commented:
> I found a problem when testing on windows. The first query enumerated is 258, and the call to `glGetPerfQueryInfoINTEL` to get the number of counters fires this error:
> 
> `'Error has been generated. GL error GL_INVALID_VALUE in GetPerfQueryInfoINTEL: (ID: 1585457696) Generic error'`

Oh right, that's one of the "internal" query that is used to plug metrics discovery https://github.com/intel/metrics-discovery
On the Windows driver they just don't report anything about it. On the Mesa driver we took care to make the raw data counters appear.
I should probably blacklist it like the TestOa as this is pretty much useless for Renderdoc.

> 
> And then you end up with `nCounters` being uninitialised. On a debug build at least this leads to iterating up to 0xcccccccc or so and looping forever (and each `glGetPerfQueryInfoINTEL` for the counters then prints an error, leading to log spam).
> 
> I think at least any variables fetched by `glGet*INTEL` should be initialised to 0 if the function might not write to them, and also it'd be good to understand what's going wrong here.

baldurk commented:
No problem! I was just checking it was still on your backlog, please don't feel pressured.

That solution sounds fine to me.

djdeath commented:
Should be good now, unfortunately I couldn't get rid of the `GL_INVALID_VALUE` when calling `GetPerfQueryInfoINTEL`.

baldurk commented:
Sorry for the delayed response, I was in the middle of something else and I wanted to finish it before context-swapping back to this.

I tested this again on windows, and while it doesn't hit any problems (aside from the GL_INVALID_VALUE message you mentioned), it also doesn't enumerate any counters at all.

`glGetFirstPerfQueryIdINTEL` returns 258, which hits the new error condition and returns early without enumerating any counters, then `glGetNextPerfQueryIdINTEL` returns 526 which also hits the same error, and then `glGetNextPerfQueryIdINTEL` returns 0 which ends the loop.

Is this intended behaviour? It's not necessarily a problem but as it stands it doesn't sound like this extension will work at all on Windows so I'd prefer to make it linux-only if that's the only place it's intended to work.

djdeath commented:
Okay, I couldn't really tell from reading the Windows source code what would be returned.
Sounds like most of it is disabled as part of the build options.

That's fine to me to limit to Linux. Do you want me to put the counter listing in `gl_counters.cpp` conditional to `ENABLED(RDOC_LINUX)` ?

baldurk commented:
OK that's fine, thanks. I think you can take it out of the case in `gl_replay.cpp` so that `m_pIntelCounters` only ever gets initialised on linux. i.e. the `if(isMesa)` case stays the same, but the `else` still only has AMD. Then `countersIntel` will always be `NULL` on windows for now. That way if in future we can add that functionality it only has to change there as well.

djdeath commented:
Thanks, changed as you suggested.

Trass3r commented on renderdoc/driver/ihv/intel/intel_gl_counters.cpp:
```diff
@@ -0,0 +1,318 @@
+  query.name.resize(len);
+  GLuint nCounters = 0;
+  GL.glGetPerfQueryInfoINTEL(queryId, (GLuint)query.name.size(), &query.name[0], &query.size,
+                             &nCounters, NULL, NULL);
```
These arguments must not be NULL. This is the reason for those `GL_INVALID_VALUE`s on Windows.

baldurk replied:
Do not make comments on PRs that were merged and resolved over a year ago.
